--- a/arch/arm/mach-gemini/board-nas4220b.c	2012-12-25 13:28:42.538603991 +0100
+++ b/arch/arm/mach-gemini/board-nas4220b.c	2012-12-25 13:01:05.976650064 +0100
@@ -26,8 +26,7 @@
 #include <asm/mach/arch.h>
 #include <asm/mach/time.h>
 
-#include <mach/hardware.h>
-#include <mach/global_reg.h>
+#include <mach/system.h>
 #include <linux/platform_data/gemini_gmac.h>
 
 #include "common.h"
@@ -203,4 +202,5 @@
 	.timer		= &ib4220b_timer,
 	.fixup          = ib4220b_fixup,
 	.init_machine	= ib4220b_init,
+	.restart	= arch_reset,
 MACHINE_END
