--- eglibc/libc/configure.in.orig	2012-01-25 07:35:24.019155048 +0100
+++ eglibc/libc/configure.in	2012-01-25 07:35:41.998139177 +0100
@@ -2152,9 +2152,6 @@
   libc_cv_gcc_builtin_expect=no
 fi
 rm -f conftest*])
-if test "$libc_cv_gcc_builtin_expect" = no; then
-  AC_MSG_ERROR([support for __builtin_expect needed])
-fi
 
 AC_CACHE_CHECK(for __builtin_memset, libc_cv_gcc_builtin_memset, [dnl
 cat > conftest.c <<\EOF
