--- linux-3.2.2/arch/arm/mach-gemini/Makefile.orig	2012-01-30 15:24:19.620851683 +0100
+++ linux-3.2.2/arch/arm/mach-gemini/Makefile	2012-01-30 15:24:45.274853575 +0100
@@ -7,6 +7,7 @@
 obj-y			:= irq.o mm.o time.o devices.o gpio.o
 
 # Board-specific support
+obj-$(CONFIG_MACH_NAS4210B)	+= board-nas4210b.o
 obj-$(CONFIG_MACH_NAS4220B)	+= board-nas4220b.o
 obj-$(CONFIG_MACH_RUT100)	+= board-rut1xx.o
 obj-$(CONFIG_MACH_WBD111)	+= board-wbd111.o
