--- a/drivers/target/target_core_configfs.c	2013-09-16 21:08:51.212675429 +0200
+++ b/drivers/target/target_core_configfs.c	2013-09-16 21:09:49.291476979 +0200
@@ -163,7 +163,7 @@
 			 *
 			 * mkdir -p $CONFIGFS/target/iscsi
 			 */
-			ret = request_module("iscsi_target_mod");
+		ret = 0;
 			if (ret < 0) {
 				pr_err("request_module() failed for"
 				       " iscsi_target_mod.ko: %d\n", ret);
@@ -176,7 +176,7 @@
 			 *
 			 * mkdir -p $CONFIGFS/target/loopback
 			 */
-			ret = request_module("tcm_loop");
+			ret = 0;
 			if (ret < 0) {
 				pr_err("request_module() failed for"
 				       " tcm_loop.ko: %d\n", ret);
