--- a/source3/passdb/secrets.c	2012-08-18 01:59:39.251910842 +0200
+++ b/source3/passdb/secrets.c	2012-08-18 01:59:51.348560121 +0200
@@ -347,7 +347,7 @@
 
 		data=(char *)secrets_fetch(old_style_key, &size);
 		if ((data == NULL) || (size < sizeof(old_style_pw))) {
-			DEBUG(0,("fetch_ldap_pw: neither ldap secret retrieved!\n"));
+			DEBUG(1,("fetch_ldap_pw: neither ldap secret retrieved!\n"));
 			SAFE_FREE(old_style_key);
 			SAFE_FREE(*dn);
 			SAFE_FREE(data);
