--- iptables-1.4.10/extensions/GNUmakefile.in.orig	2012-04-18 20:32:23.000000000 +0200
+++ iptables-1.4.10/extensions/GNUmakefile.in	2012-04-18 20:32:55.000000000 +0200
@@ -82,7 +82,7 @@
 
 install: ${targets_install}
 	@mkdir -p "${DESTDIR}${xtlibdir}";
-	if test -n "${targets_install}"; then install -pm0755 $^ "${DESTDIR}${xtlibdir}/"; fi;
+	if test "   " -ne "${targets_install}"; then install -pm0755 $^ "${DESTDIR}${xtlibdir}/"; fi;
 
 clean:
 	rm -f *.o *.oo *.so *.a {matches,targets}[46].man initext4.c initext6.c;
